### PR TITLE
[Do not merge] Addition of CRUD operations for Application Insights Components & Synthetic Monitors

### PR DESCRIPTION
**Pull request by d3r3kk**
Add the swagger spec for all CRUD operations to do with Components and Synthetic Monitors (web tests).

Spawned from PR "Application Insights: CRUD for a new Instance #757" originally put forth by @tombuildsstuff and worked on by @amarzavery.

Please review and let me know if you have any concerns: @SergeyKanzhelev, @salameer. Any feedback is greatly appreciated.

----
### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [x] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
@d3r3kk this is great - thanks!

**Comment by d3r3kk** [comment]
@tombuildsstuff Please have a look and let me know if there are any problems with what you were attempting to do with the API. Having real-world usage against this API will be very helpful feedback!

**Comment by tombuildsstuff** [comment]
@d3r3kk thanks for this - to keep you in the loop, I took a quick look but can't generate the Go SDK at this point from the composite swagger right now due to an error:

```
 $ dotnet AutoRest.dll -Namespace applicationinsights -CodeGenerator Go -Modeler Swagger -Input ~/code/src/github.com/Azure/azure-rest-api-specs/arm-insights/compositeInsightsManagementClient.json
ERROR: Operations API must be implemented for the service.
	Path: ~/code/src/github.com/Azure/azure-rest-api-specs/arm-insights/compositeInsightsManagementClient.json#/paths
WARNING: Parameters "subscriptionId" or "api-version" are referenced but not defined in the parameters section of Service Definition
	Path: ~/code/src/github.com/Azure/azure-rest-api-specs/arm-insights/compositeInsightsManagementClient.json#/parameters
FATAL: Invalid swagger 2.0 specification. Missing version property.
FATAL: Error generating client model: AutoRest.Core.Logging.CodeGenerationException: Invalid swagger 2.0 specification. Missing version property.
   at AutoRest.Swagger.SwaggerModeler.InitializeClientModel() in ~/code/src/github.com/Azure/AutoRest/src/modeler/AutoRest.Swagger/SwaggerModeler.cs:line 226
   at AutoRest.Swagger.SwaggerModeler.Build(ServiceDefinition serviceDefinition) in ~/code/src/github.com/Azure/AutoRest/src/modeler/AutoRest.Swagger/SwaggerModeler.cs:line 93
   at AutoRest.Swagger.SwaggerModeler.Build() in ~/code/src/github.com/Azure/AutoRest/src/modeler/AutoRest.Swagger/SwaggerModeler.cs:line 72
   at AutoRest.Core.AutoRestController.Generate() in ~/code/src/github.com/Azure/AutoRest/src/core/AutoRest.Core/AutoRestController.cs:line 57
ERROR: Error generating client model: AutoRest.Core.Logging.CodeGenerationException: Invalid swagger 2.0 specification. Missing version property.
   at AutoRest.Swagger.SwaggerModeler.InitializeClientModel() in ~/code/src/github.com/Azure/AutoRest/src/modeler/AutoRest.Swagger/SwaggerModeler.cs:line 226
   at AutoRest.Swagger.SwaggerModeler.Build(ServiceDefinition serviceDefinition) in ~/code/src/github.com/Azure/AutoRest/src/modeler/AutoRest.Swagger/SwaggerModeler.cs:line 93
   at AutoRest.Swagger.SwaggerModeler.Build() in ~/code/src/github.com/Azure/AutoRest/src/modeler/AutoRest.Swagger/SwaggerModeler.cs:line 72
   at AutoRest.Core.AutoRestController.Generate() in ~/code/src/github.com/Azure/AutoRest/src/core/AutoRest.Core/AutoRestController.cs:line 57
ERROR: Code generation failed.
ERROR: AutoRest.dll -Namespace applicationinsights -CodeGenerator Go -Modeler Swagger -Input ~/code/src/github.com/Azure/azure-rest-api-specs/arm-insights/compositeInsightsManagementClient.json
```

I [noticed some errors in Travis](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/207471208#L1673) too - but that the Go generator isn't hooked up to that yet?

File: arm-insights/2015-05-01/swagger/ApplicationInsights_SyntheticMonitor.json [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+      "required": true,
+      "type": "string",
+      "description": "The Azure subscription Id."
+    },
+    "ApiVersionParameter": {
+      "name": "api-version",
+      "in": "query",
+      "required": true,
+      "type": "string",
+      "description": "Client API version."
```
Comment (olydis) [inline arm-insights/2015-05-01/swagger/ApplicationInsights_SyntheticMonitor.json]: All documents of composite Swagger need to agree about parameters defined on the global scope 100%. In this case, AutoRest complains about a mismatching description ;-) In another document in here I see (note capitalization):
```
    "ApiVersionParameter": {
      "name": "api-version",
      "in": "query",
      "required": true,
      "type": "string",
      "description": "Client Api Version."
    },
```
Please make sure all documents use precisely the same `ApiVersionParameter` description.

**Comment by amarzavery** [comment]
@tombuildsstuff - Those are validation errors and not codegen errors. Both use the same logger. If an error was logged ever the final exit code is 1 and you see that error message. We are fixing that part.
However, the code should still be generated. Please check the output directory, you should see the generated client in there. Sorry for the bad experience.

File: arm-insights/2015-05-01/swagger/ApplicationInsights_ComponentClient.json [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "A list containing 0 or more Application Insights component definitions.",
+            "schema": {
+              "$ref": "#/definitions/ApplicationInsightsComponentListResult"
+            },
+            "examples": {
```
Comment (amarzavery) [inline arm-insights/2015-05-01/swagger/ApplicationInsights_ComponentClient.json]: instead of providing example in the spec, can you please provide them using the "x-ms-examples" extension. 
Take a look at the RedisCache swagger spec as an example over [here](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-redis/2016-04-01/swagger/redis.json#L43).

**Review by amarzavery — CHANGES_REQUESTED** [review]

File: arm-insights/2015-05-01/swagger/ApplicationInsights_SyntheticMonitor.json (hunk 1 above)
Comment (d3r3kk) [inline arm-insights/2015-05-01/swagger/ApplicationInsights_SyntheticMonitor.json]: @olydis Fixed in my latest commit - I was unfortunately using the Nuget AutoRest package which is currently at v0.17.0 and did not report this error. I've since updated to use the NPM autorest package which is at 1.0.1 and have corrected the problem.

@tombuildsstuff I've just now built the Go API and it succeeded without warnings.

File: arm-insights/2015-05-01/swagger/ApplicationInsights_ComponentClient.json (hunk 2 above)
Comment (d3r3kk) [inline arm-insights/2015-05-01/swagger/ApplicationInsights_ComponentClient.json]: Looking...

**Comment by gucalder** [comment]
This change is being applied to the arm-insights folder, which is not the AppInsights folder.

**Comment by hovsepm** [comment]
@d3r3kk, @amarzavery, @olydis 

This PR targets wrong folder. Please re-target it to the correct folder per @gucalder comment.

File: arm-insights/2015-05-01/swagger/ApplicationInsights_ComponentClient.json (hunk 2 above)
Comment (d3r3kk) [inline arm-insights/2015-05-01/swagger/ApplicationInsights_ComponentClient.json]: Complete, have a look and let me know - tested using 'npm test -- test/model.js' and got that running cleanly for the Components and SyntheticMonitor specs I've added only.

**Comment by d3r3kk** [comment]
@gucalder, @hovsepm, I was under the understanding that Insights _includes_ Application Inisghts. Indeed, the 'Alerts' API that is found under ../arm-insights/ is provided by the same web service that provides Component and Synthetic Monitor APIs. I can create an arm-appinsights folder though, I have no problem doing so.

**Comment by d3r3kk** [comment]
@amarzavery I've changed all examples to using the x-ms-examples pattern.

**Comment by amarzavery** [comment]
@d3r3kk @hovsepm 
I think there is some miscommunication over here. When @tombuildsstuff (our customer) sent the initial PR I was trying to find the right service team internally who can review the PR as they would know the APIs in depth. In that process, I approached @gucalder and his manager @andyshen since the provider namespace was **Microsoft.Insights**. They told me that even though ApplicationInsights and Insights share the same provider namespace (/subscriptions/{subscriptionId}/providers/ **Microsoft.Insights**/ ...), they are completely different services with nothing in common. To avoid this confusion Insights is changing (just the folder name and client library (package name) to Monitoring. I think the swagger spec in this PR should reside in a separate folder named `arm-applicationinsights`.

Does that make sense?

**Comment by amarzavery** [comment]
@d3r3kk - I would suggest syncing internally with @hovsepm @gucalder and @andyshen and then make the final push. This will help us reach to the logical confusion faster.

**Comment by d3r3kk** [comment]
@amarzavery I've actually been chatting with @hovsepm and @gucalder in an email thread I'll add you to this morning. *Apologies for any confusion I may have caused here*!

I am trying to determine the best possible outcome for the customer (components, synthetic monitors, and alerts, are all coupled and will be used in tandem by our users when they use application insights APIs).

**Comment by d3r3kk** [comment]
Not sure how to resolve the build failure with the Azure.Python step. I cannot reproduce the failure on my local machine at all. Seems to be failing with the 'update client' part of the SwaggerToApi.py script... If someone tells me how I can resolve that I will gladly take care of it.

**Comment by d3r3kk** [comment]
@amarzavery and @gucalder is there anything else lacking in this PR that I can fix up so that it gets integrated?

**Review by amarzavery — APPROVED** [review]

**Comment by amarzavery** [comment]
Sorry I approved by mistake. Please ignore the approval

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+      }
+    },
+    "WebTestProperties": {
+      "description": "Metadata describing a synthetic monitor for an Azure resource.",
+      "required": [
+        "Name",
+        "Kind"
+      ],
+      "properties": {
+        "SyntheticMonitorId": {
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: This is a general comment for all the model properties in swagger.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+          "enum": [
+            "ping",
+            "multistep"
+          ],
+          "x-ms-enum": {
+            "name": "WebTestKind",
+            "modelAsString": false
+          }
+        },
+        "RetryEnabled": {
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: is this a required or optional property? If this is optional then what is the default value?

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+          "x-ms-client-name": "WebTestName",
+          "type": "string",
+          "description": "User defined name if this WebTest."
+        },
+        "Description": {
+          "type": "string",
+          "description": "Purpose/user defined descriptive test for this WebTest."
+        },
+        "Enabled": {
+          "type": "boolean",
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: the description says that the default is false. Please add the default to the property as well.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+          "description": "Current state of this component, whether or not is has been provisioned within the resource group it is defined. Users cannot change this value but are able to read from it. Values will include Succeeded, Deploying, Canceled, and Failed.",
+          "readOnly": true
+        }
+      }
+    },
+    "WebTestGeolocation": {
+      "description": "Geo-physical location to run a web test from.",
+      "type": "object",
+      "properties": {
+        "Id": {
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: This doesn't make sense to me. The model name is talking about some GeoLocation and there is only one property named id. Does the id provide the location? Shouldn't there be a property named location or something equivalent?

Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - is this a required property?
- is this readOnly?

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+          "description": "Purpose/user defined descriptive test for this WebTest."
+        },
+        "Enabled": {
+          "type": "boolean",
+          "description": "Is the test actively being monitored, default is false."
+        },
+        "Frequency": {
+          "type": "integer",
+          "format": "int32",
+          "description": "Interval between test runs for this WebTest."
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: is the user does not provide any value then what will the server do? Assume some value? If yes, then the default should be documented

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+        },
+        "Timeout": {
+          "type": "integer",
+          "format": "int32",
+          "description": "Seconds until this WebTest will timeout and fail."
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: same as above

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+          "items": {
+            "$ref": "#/definitions/ApplicationInsightsComponent"
+          }
+        },
+        "nextLink": {
+          "type": "string",
+          "description": "The URI to get the next set of Application Insights component defintions if too many components where returned in the result set."
+        }
+      },
+      "x-ms-azure-resource": true
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: Please remove x-ms-azure-resource: true extension from this model. It is not required over here.

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+          "description": "The unique ID of your application."
+        },
+        "AppId": {
+          "type": "string",
+          "readOnly": true,
+          "description": "Application Insights Unique ID for your Application."
+        },
+        "Application_Type": {
+          "type": "string",
+          "description": "Type of application being monitored. Example values: Web, other. Cannot be changed after the component has been created."
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: `"x-ms-mutability": ["create", "read"]`

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+        },
+        "Flow_Type": {
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: > "Used by the Application Insights system to determine what kind of flow this component was created by. Users should set this to 'Bluefield' when creating a component via the REST API."

So,
- is this updatable? If not then can you add that in the description? Also add `"x-ms-mutability": ["create", "read"]`.
- if it is updatable then what are the other values apart from `'Bluefield'`
- should this be an enum?

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+      ],
+      "required": [
+        "kind"
+      ],
+      "description": "An Application Insights component definition."
+    },
+    "ApplicationInsightsComponentProperties": {
+      "description": "Properties that define an Application Insights component resource.",
+      "properties": {
+        "ApplicationId": {
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: should the user provide this or it is readOnly?

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 10 above)
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - Should this be an enum?
- Is Web the only value that the user can provide?

**Review by amarzavery — CHANGES_REQUESTED** [review]

**Comment by veronicagg** [comment]
@amarzavery I'm assigning this one to you, since it looks like you've been reviewing it, if you need someone else to help, feel free to ask. thanks!

**Comment by d3r3kk** [comment]
Awesome feedback, thanks @amarzavery! Sorry I haven't responded for 5 days, I've since moved onto other efforts for my team, but will revisit this PR by this coming Friday, 24th of March.

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+      "type": "oauth2",
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "flow": "implicit",
+      "description": "Azure Active Directory OAuth2 Flow.",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: Done. Added in my upcoming change - created a new file 'aiOperations_API.json' as this seems unrelated.

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+    "/subscriptions/{subscriptionId}/providers/microsoft.insights/components": {
+      "get": {
+        "description": "Gets a list of all Application Insights components within a subscription.",
+        "operationId": "Component_List",
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - [x] **Done**. Both for Component**s** and WebTest**s**

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+        },
+        "x-ms-examples": {
+          "ComponentListByResourceGroup": { "$ref": "../examples/ComponentListByResourceGroup.json" }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/microsoft.insights/components/{resourceName}": {
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - [x] **Done**. Added Patch operation and simplified tags-only resource for specifying it.

File: arm-appinsights/2015-05-01/swagger/component_API.json [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+  "definitions": {
+    "Resource": {
+      "properties": {
+        "id": {
+          "type": "string",
+          "readOnly": true,
+          "description": "Azure resource Id"
+        },
+        "name": {
+          "type": "string",
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - [x] **Done**

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 12 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - [x] **Done** I have made this readOnly. This field is tied to the 'Name'  field and cannot be changed directly.

This seems like a deprecated field that my team has yet to remove from the API. Perhaps I should just cull it from the spec altogether?

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 11 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: Great question. 

Ideally, we would like this field to always be a specific value when used from the REST API via swagger. Is there a way I can 'hard-code' this field to Bluefield, and leave it that way no matter how the user sets it?

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 9 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - [x] **Done**, copy-paste error.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/microsoft.insights/webtests": {
+      "get": {
+        "description": "Get all Application Insights synthetic monitors defined within a specified resource group.",
+        "operationId": "SyntheticMonitor_ListByResourceGroup",
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] **Done** I've completely removed the use of the term 'Synthetic Monitor' with my upcoming change. Synthetic monitor is a term we use throughout our documentation, but WebTests is the name of the resource. I believe my teammates have some thoughts on this subject and if they can come up with a compelling reason for me to change it back I will do so.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+          "SyntheticMonitorListByResourceGroup": {
+            "$ref": "../examples/SyntheticMonitorListByResourceGroup.json"
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/microsoft.insights/webtests/{webTestName}": {
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] **Done**

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/WebTestNameParameter"
+          },
+          {
+            "name": "WebTestDefinition",
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] **Done** as stated above, I've wholesale renamed the Synthetic Monitor API back to WebTests. (Including the file name!)

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+  "definitions": {
+    "Resource": {
+      "properties": {
+        "id": {
+          "type": "string",
+          "readOnly": true,
+          "description": "Azure resource Id"
+        },
+        "name": {
+          "type": "string",
```
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] **Done**

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 5 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] **Done**. I've removed the suggestion of a default state from the description.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 6 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] **Done** Agreed, this was confusing. Apart from the 'description' being nonsense, then naming of this particular property is confusing to users. I've left it as it is required to be on the wire, but added `"x-ms-client-name": "location"` to hopefully better clarify the purpose of this item.

Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: This is an editable, required property. Have updated the description to help with this.

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 10 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - [x] **Done** Yes I believe that an enum in this context makes a great deal of sense. I've altered the definition to be such as well as to add the mutability decoration.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 6 above)
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: I hope you specified "required" correctly the way it should be specified as per the swagger specification. For example take a look over [here](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-compute/2016-04-30-preview/swagger/compute.json#L5367) , how location is described as a required property of the Resource..

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 12 above)
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: Swagger should correctly and completely represent the wire format for a given api-version. No exceptions to this rule. When the service removes it and does not send it on the wire then this should be removed, otherwise not.

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 11 above)
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: There is a way to make a model property or a parameter a constant. You should mark it as "required" and add an enum constraint `"enum": ["Bluefield"]` with one value. Autorest will treat this as a constant and will not expose it. It will always serialize that property/parameter with the specified value.

One thing that I am confused about, from your comment
> and leave it that way **no matter how the user sets it?**

If you make something a constant the way I specified above, then the user will never be able to set it. It will always be "Bluefield" on the wire.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 17 above)
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: This will cause linter violations. We have a recommendation from the ARM team to name resource types the same as they are in the Url. Different names causes confusion.
@ravbhatnagar @salameer @veronicagg - please take a note of this.

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 12 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: Gotcha, will leave it in place and describe it as such.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 19 above)
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: Ah I see. From the above comment, it seemed that you want to keep the name synthetic monitoring as that is what your team mates think.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 5 above)
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: then is this an optional property or a required property? If require then this needs to be described in swagger accordingly.

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 11 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: This is very helpful, and is useful when creating new Components, they do indeed all get created with the 'Bluefield' value. However, when I get a list of components, I get components that have different FlowTypes (ie. 'Redfield'). In the C# SDK that I generate from my swagger, I now get deserialization exceptions.

Apologies for the confusion, perhaps this will help?

My question then changes to: 
Can I specify in my swagger spec a way to ensure that FlowType is always *set* to 'Bluefield' but that still allows the *read* of other values of FlowType. It is always acceptable for us to set the field to 'Bluefield' if we ever update a component that had a different value previously.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 5 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: This is an optional property.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 7 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] Done, the default value for Timeout and Frequency are now reflected in the spec.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 8 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] Done (as above)

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 4 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: Optional, no default value. (If left unset then the server will leave it unset)

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 6 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] Done

File: arm-appinsights/2015-05-01/swagger/component_API.json (hunk 11 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/component_API.json]: - [x] Done. Set as an enum field, with a default value, and set the enum type 'modelAsString' to true.

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 17 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] Done

File: arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json (hunk 3 above)
Comment (d3r3kk) [inline arm-appinsights/2015-05-01/swagger/syntheticMonitor_API.json]: - [x] Done. I took some fiddler traces of how data is sent back from the server and made sure each field is written exactly as it appears on the wire.

**Comment by d3r3kk** [comment]
Thanks for your patience everyone. I've posted my changes requested by @amarzavery along with some new example files as well.

Please review and let me know if there is any further work to do to ensure the quality of our spec is inline with the repo.

Thanks!

**Comment by amarzavery** [comment]
@d3r3kk - The model validator found issues. Please take a look at the Travis CI log over here https://travis-ci.org/Azure/azure-rest-api-specs/jobs/216112575#L1042

File: test/linter.js [hunk 21]
```diff
@@ -10,9 +10,9 @@ describe('AutoRest Linter validation:', function () {
   let swaggersToProcess = utils.getFilesChangedInPR();
   // Useful when debugging a test for a particular swagger. 
   // Just update the regex. That will return an array of filtered items.
-  // swaggersToProcess = swaggersToProcess.filter(function(item) {
-  //   return (item.match(/.arm-containerregistry.*2017-03-01.*/ig) !== null);
-  // });
+  swaggersToProcess = swaggersToProcess.filter(function(item) {
```
Comment (amarzavery) [inline test/linter.js]: Please revert this change.

File: arm-appinsights/2015-05-01/swagger/webTests_API.json [hunk 22]
```diff
@@ -359,12 +361,14 @@
...
           "format": "int32",
-          "description": "Interval between test runs for this WebTest."
+          "description": "Interval in seconds between test runs for this WebTest. Default value is 300.",
+          "default": 300
         },
         "Timeout": {
           "type": "integer",
           "format": "int32",
-          "description": "Seconds until this WebTest will timeout and fail."
+          "description": "Seconds until this WebTest will timeout and fail. Default valu is 30.",
```
Comment (amarzavery) [inline arm-appinsights/2015-05-01/swagger/webTests_API.json]: Default valu is 30.", -> missing 'e' in value

**Comment by amarzavery** [comment]
- There is an enum mismatch in the examples and what is specified in swagger. Values you specify in the enum array in the swagger spec are case sensitive. This also affects deserialization in the generated sdks. Make sure that the casing of all the enum values matches exactly what is being sent on the wire.

-  x-ms-example "webTestUpdate" in operation "WebTests_CreateOrUpdate". 'OBJECT_MISSING_REQUIRED_PROPERTY', -> 'SyntheticMonitorId'

**Comment by d3r3kk** [comment]
@amarzavery I believe I've rectified the issues with examples / typos. Apologies for checking in the linter test.

**Comment by amarzavery** [comment]
Hey @d3r3kk the travis ci shows that there are still a bunch of model validation errors:
 https://travis-ci.org/Azure/azure-rest-api-specs/jobs/216467585#L1046

- Can you please fix all of those?

**Comment by d3r3kk** [comment]
- [x] **Done**.

What about the insights/operations link. I see that I fail in the linter script for those for the two main spec files, but I do not fail in the single file I included that in.

It seems I have to include that path in each of my swagger spec files to pass the linter. Would that be preferable to the solution I've devised?

**Comment by amarzavery** [comment]
@d3r3kk - The linter is run on each swagger spec separately. It has no ideas that the swagger is a part of the composite swagger. Hence please ignore the operations API Implementation error in other specs. If you have included the API in one of the swagger specs then you are good to go.

**Review by amarzavery — APPROVED** [review]

**Comment by amarzavery** [comment]
Thanks @d3r3kk for authoring the swagger and incorporating all the feedback. 

@tombuildsstuff - Sorry for the delay. But you should be able to use this (applicationInsights) swagger and the clients generated from this swagger.  Please feel free to file issues if you see any problems. Thanks for the initial PR with swagger. That helped us deliver this today.

**Comment by salameer** [comment]
@d3r3kk  Thanks again, @tombuildsstuff, we'll make sure that this get's out into our next Go SDK  release. adding @marstr for visibility
